storage/remote/queue_manager.go

```diff
@@ -1055,16 +1061,33 @@ func sendWriteRequestWithBackoff(ctx context.Context, cfg config.QueueConfig, s
 			return nil
 		}
 
+		if hasPreviouslyRetried && maxRetries > 0 && numRetries > maxRetries {
+			return errors.Errorf("retries count exceeded max retries: num-retries: %d max-retries: %d", numRetries, maxRetries)
+		}
+
 		// If the error is unrecoverable, we should not retry.
-		if _, ok := err.(RecoverableError); !ok {
+		backoffErr, ok := err.(RecoverableError)
+		if !ok {
 			return err
 		}
 
+		if !hasPreviouslyRetried {
+			// Once we set the sleepDuration, we do not change it even if the Retry-After header is changed in subsequent responses.
+			sleepDuration = backoff
+			if backoffErr.isRateLimiting && sleepDuration < backoffErr.retryAfter {
+				sleepDuration = backoffErr.retryAfter
+			}
+			if backoffErr.maxRetries > 0 {
+				maxRetries = backoffErr.maxRetries
+			}
+		}
+		hasPreviouslyRetried = true
+
 		// If we make it this far, we've encountered a recoverable error and will retry.
 		onRetry()
-		level.Debug(l).Log("msg", "failed to send batch, retrying", "err", err)
+		level.Debug(l).Log("msg", "failed to send batch, retrying", "err", backoffErr.Error())
```

Comment:
Somewhat a reminder for me in case a rebase doesn't go well, this should be `Warn` level, and failed should be Failed.

Reply:
I have implemented this in the recent push.